go/cmd/dolt/commands/diff.go

```diff
@@ -651,30 +689,68 @@ func getColumnNamesString(fromSch, toSch schema.Schema) string {
 	return strings.Join(cols, ",")
 }
 
-func writeDiffResults(
-	ctx *sql.Context,
+func getDiffRows(
+	sqlCtx *sql.Context,
 	diffQuerySch sql.Schema,
-	targetSch sql.Schema,
-	iter sql.RowIter,
-	writer diff.SqlRowDiffWriter,
-) error {
-	ds, err := newDiffSplitter(diffQuerySch, targetSch)
+	rowIter sql.RowIter,
+	unionSch sql.Schema,
+	filterChangedCols bool,
+) ([]rowDiff, []rowDiff, map[string]bool, error) {
+	oldRows, newRows := []rowDiff{}, []rowDiff{}
+	modifiedColNames := make(map[string]bool)
+
+	ds, err := newDiffSplitter(diffQuerySch, unionSch)
 	if err != nil {
-		return err
+		return oldRows, newRows, modifiedColNames, err
 	}
 
 	for {
-		r, err := iter.Next(ctx)
+		r, err := rowIter.Next(sqlCtx)
 		if err == io.EOF {
-			return writer.Close(ctx)
-		} else if err != nil {
-			return err
+			break
 		}
 
 		oldRow, newRow, err := ds.splitDiffResultRow(r)
 		if err != nil {
-			return err
+			return oldRows, newRows, modifiedColNames, err
+		}
+
+		if filterChangedCols {
+			var filteredOldRow, filteredNewRow rowDiff
+			for i, changeType := range oldRow.colDiffs {
+				if changeType != diff.None || unionSch[i].PrimaryKey {
+					modifiedColNames[unionSch[i].Name] = true
+
+					filteredOldRow.row = append(filteredOldRow.row, oldRow.row[i])
+					filteredOldRow.colDiffs = append(filteredOldRow.colDiffs, oldRow.colDiffs[i])
+					filteredOldRow.rowDiff = oldRow.rowDiff
+
+					filteredNewRow.row = append(filteredNewRow.row, newRow.row[i])
+					filteredNewRow.colDiffs = append(filteredNewRow.colDiffs, newRow.colDiffs[i])
+					filteredNewRow.rowDiff = newRow.rowDiff
+				}
+			}
+
+			oldRows = append(oldRows, filteredOldRow)
```

Comment:
How hard would it be to keep this code working from an iterator and not having to load all diff results into memory at once? 

I think it would be ideal if the code could traverse the iterator once to figure out the changed columns and the right schema for the filtered view, then reset the iterator to traverse again and apply any filtering. The benefit is of course just that we don't have to load all the diff results into memory. That's fine for small diffs, but we do have to deal with very large diffs as well, and it seems like we could tweak this code just a little more to be more efficient there. Let me know what you think.

Reply:
That makes sense! I'll implement the changes in a follow-up commit.